crates/core/tedge_api/src/plugin.rs

```diff
@@ -0,0 +1,619 @@
+/// thin-edge.
+///
+/// # Note
+///
+/// As a plugin author, you will not have to implement this trait.
+/// The core of thin-edge will use this trait to hand over an object to a plugin that can then be
+/// used to communicate with other plugins (as described above).
+///
+pub trait PluginDirectory: Send + Sync {
+    /// Request an `Address` object for a given plugin which can receive messages of a type
+    /// included in the message bundle `MB`.
+    ///
+    /// ## Also see
+    ///
+    /// - [`make_message_bundle`] On how to define your own named message bundle
+    fn get_address_for<RB: ReceiverBundle>(&self, name: &str) -> Result<Address<RB>, DirectoryError>;
+
+    /// Request an `Address` to the core itself. It will only accept messages from the
+    /// [`CoreMessages`] bundle.
+    fn get_address_for_core(&self) -> Address<CoreMessages>;
+}
+
+/// The plugin configuration as a `toml::Spanned` table.
+///
+/// It is important that configuration errors are communicated precisely
+/// and concisely. Reporting the span is not a must, but greatly helps users
+/// in diagnostics of errors as well as sources of configuration.
+pub type PluginConfiguration = toml::Spanned<toml::value::Value>;
+
+/// A plugin builder for a given plugin
+///
+/// A type implementing PluginBuilder is used by the core of thin-edge to instantiate a plugin
+/// implementation.
+///
+/// # Note
+///
+/// Plugin authors want to implement this trait so that the core of thin-edge can instantiate their
+/// plugin if the configuration of thin-edge desires so.
+///
+/// The implementation of the trait is then used by thin-edge to verify that the configuration
+/// passed to the plugin is sound (what "sound" means in this context is to be decided by the
+/// plugin author, i.e. the author of the implementation of this trait).
+///
+/// The plugin author must also name all message types the plugin which is about to be instantiated
+/// can receive (see `PluginBuilder::kind_message_types`).
+#[async_trait]
+pub trait PluginBuilder<PD: PluginDirectory>: Sync + Send + 'static {
+    /// The name for the kind of plugins this creates, this should be unique and will prevent startup otherwise
+    ///
+    /// The "kind name" of a plugin is used by the configuration to name what plugin is to be
+    /// instantiated. For example, if the configuration asks thin-edge to instantiate a plugin
+    /// of kind "foo", but only a plugin implementation of kind "bar" is compiled into thin-edge,
+    /// the software is able to report misconfiguration on startup.
+    fn kind_name() -> &'static str
+    where
+        Self: Sized;
+
+    /// A list of message types the plugin this builder creates supports
+    ///
+    /// This function must return a `HandleTypes` object which represents the types of messages
+    /// that a plugin is able to handle.
+    ///
+    /// To create an instance of this type, you must use the [`HandleTypes::get_handlers_for`] method.
```

Comment:
A note for `HandleTypes::empty` should be added

Reply:
Although that method can be removed and instead a note could be given that `()` is an acceptable `MessageBundle`.